rust/main/chains/hyperlane-ton/src/client/provider.rs

```diff
@@ -0,0 +1,773 @@
+    h512_to_bytes, BlockInfo, ChainCommunicationError, ChainInfo, ChainResult, FixedPointNumber,
+    HyperlaneChain, HyperlaneCustomErrorWrapper, HyperlaneDomain, HyperlaneProvider, TxOutcome,
+    TxnInfo, TxnReceiptInfo, H256, H512, U256,
+};
+use reqwest::{Client, Response};
+use serde_json::{json, Value};
+use tokio::time::sleep;
+
+use crate::client::error::CustomHyperlaneError;
+use crate::run_get_method::StackItem;
+use crate::{
+    trait_builder::TonConnectionConf,
+    traits::ton_api_center::TonApiCenter,
+    types::{
+        account_state::AccountStateResponse, block_response::BlockResponse,
+        message::MessageResponse, message::SendMessageResponse,
+        run_get_method::RunGetMethodResponse, transaction::TransactionResponse,
+        wallet_state::WalletStatesResponse,
+    },
+    utils::conversion::ConversionUtils,
+};
+use tonlib_core::TonAddress;
+use url::Url;
+
+#[derive(Clone, new)]
+pub struct TonProvider {
+    pub http_client: Client,
+    pub connection_conf: TonConnectionConf,
+    pub domain: HyperlaneDomain,
+}
+
+impl TonProvider {
+    async fn post_request(
+        &self,
+        url: Url,
+        params: &Value,
+    ) -> Result<Response, ChainCommunicationError> {
+        self.http_client
+            .post(url)
+            .header("accept", "application/json")
+            .header("Content-Type", "application/json")
+            .header("X-API-Key", self.connection_conf.api_key.clone())
+            .json(params)
+            .send()
+            .await
+            .map_err(|e| {
+                warn!("Error sending request: {:?}", e);
+                ChainCommunicationError::Other(HyperlaneCustomErrorWrapper::new(Box::new(e)))
+            })
+    }
+
+    async fn query_request<T: serde::Serialize + ?Sized>(
+        &self,
+        url: Url,
+        params: &T,
+    ) -> Result<Response, reqwest::Error> {
+        self.http_client
+            .get(url)
+            .query(params)
+            .header("accept", "application/json")
+            .header("Content-Type", "application/json")
+            .header("X-API-Key", self.connection_conf.api_key.clone())
+            .send()
+            .await
+    }
+}
+impl std::fmt::Debug for TonProvider {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TonProvider")
+            .field("client", &self.http_client)
+            .field("domain", &self.domain)
+            .finish()
+    }
+}
+impl HyperlaneChain for TonProvider {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(self.clone())
+    }
+}
+
+#[async_trait]
+impl HyperlaneProvider for TonProvider {
+    async fn get_block_by_height(&self, _height: u64) -> ChainResult<BlockInfo> {
+        let response = self
+            .get_blocks(
+                -1, // Masterchain
+                None,
+                Some(_height),
```

Comment:
Since parameter `_height` is used, it should not have leading underscore in its name.